// zeplin-extension-style-kit/values/scalar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function round(number, precision) {
  var formattedNumber = Number(number).toLocaleString("en-US", {
    useGrouping: false,
    maximumFractionDigits: precision
  });
  return Number(formattedNumber);
}

var DEFAULT_PRECISION = 2;

var Scalar = /*#__PURE__*/function () {
  function Scalar(value) {
    var precision = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : DEFAULT_PRECISION;

    _classCallCheck(this, Scalar);

    this.value = value;
    this.precision = precision;
  }

  _createClass(Scalar, [{
    key: "valueOf",
    value: function valueOf() {
      var value = this.value,
          precision = this.precision;
      return "scalar::v:".concat(value, ":p:").concat(precision);
    }
  }, {
    key: "equals",
    value: function equals(other) {
      return this.value === other.value;
    }
  }, {
    key: "toStyleValue",
    value: function toStyleValue() {
      return "".concat(round(this.value, this.precision));
    }
  }]);

  return Scalar;
}();

var _default = Scalar;
exports.default = _default;
