// zeplin-extension-style-kit/values/utility/fourDirectionalValue.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var FourDirectionalValue = /*#__PURE__*/function () {
  function FourDirectionalValue(_ref) {
    var top = _ref.top,
        right = _ref.right,
        bottom = _ref.bottom,
        left = _ref.left;

    _classCallCheck(this, FourDirectionalValue);

    this.top = top;
    this.right = right;
    this.bottom = bottom;
    this.left = left;
  }

  _createClass(FourDirectionalValue, [{
    key: "valueOf",
    value: function valueOf() {
      return "fourDirectionalValue::t:(".concat(this.top.valueOf(), "):r:(").concat(this.right.valueOf(), "):b:(").concat(this.bottom.valueOf(), "):l:(").concat(this.left.valueOf(), ")");
    }
  }, {
    key: "equals",
    value: function equals(other) {
      return this.top.equals(other.top) && this.right.equals(other.right) && this.bottom.equals(other.bottom) && this.left.equals(other.left);
    }
  }, {
    key: "toStyleValue",
    value: function toStyleValue(params) {
      if (this.top.equals(this.bottom) && this.right.equals(this.left) && this.top.equals(this.right)) {
        return this.top.toStyleValue(params);
      }

      if (this.top.equals(this.bottom) && this.right.equals(this.left)) {
        return "".concat(this.top.toStyleValue(params), " ").concat(this.right.toStyleValue(params));
      }

      if (this.right.equals(this.left)) {
        return "".concat(this.top.toStyleValue(params), " ").concat(this.right.toStyleValue(params), " ").concat(this.bottom.toStyleValue(params));
      }

      return "".concat(this.top.toStyleValue(params), " ").concat(this.right.toStyleValue(params), " ").concat(this.bottom.toStyleValue(params), " ").concat(this.left.toStyleValue(params));
    }
  }]);

  return FourDirectionalValue;
}();

var _default = FourDirectionalValue;
exports.default = _default;
