// zeplin-extension-style-kit/declarations/alignSelf.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _constants = require("../constants");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var valueMapper = {
  center: "center",
  max: "flex-end",
  min: "flex-start",
  stretch: "stretch",
  inherit: "inherit"
};

var AlignSelf = /*#__PURE__*/function () {
  function AlignSelf(value) {
    _classCallCheck(this, AlignSelf);

    this.value = value;
  }

  _createClass(AlignSelf, [{
    key: "name",
    get: function get() {
      return _constants.STYLE_PROPS.ALIGN_SELF;
    }
  }, {
    key: "equals",
    value: function equals(other) {
      return this.value === other.value;
    }
  }, {
    key: "getValue",
    value: function getValue() {
      return valueMapper[this.value];
    }
  }]);

  return AlignSelf;
}();

var _default = AlignSelf;
exports.default = _default;
